executor/adapter.go

```diff
@@ -1492,6 +1497,52 @@ func resetCTEStorageMap(se sessionctx.Context) error {
 	return nil
 }
 
+// RecordHistoryStats is used to record history stats that is more accurate than estimation
+// Currently, only support "Selection + TableScan" pattern
+func (a *ExecStmt) RecordHistoryStats() {
+	sessVars := a.Ctx.GetSessionVars()
+	if sessVars.InRestrictedSQL || !sessVars.StmtCtx.InSelectStmt {
+		return
+	}
+
+	sql := FormatSQL(a.GetTextToLog(true))()
+	stmtCtx := sessVars.StmtCtx
+	flat := getFlatPlan(stmtCtx)
+	if flat == nil || len(flat.Main) == 0 {
+		return
+	}
+
+	for _, flatOp := range flat.Main {
+		pp, isPhysicalPlan := flatOp.Origin.(plannercore.PhysicalPlan)
+		if !isPhysicalPlan {
+			continue
+		}
+		if x, isSelect := pp.(*plannercore.PhysicalSelection); isSelect {
+			if _, ok := x.Children()[0].(*plannercore.PhysicalTableScan); ok {
+				checkAndRecordHistoryStats(stmtCtx, sql, x)
+			}
+		}
+	}
+}
+
+func checkAndRecordHistoryStats(stmtCtx *stmtctx.StatementContext, sql string, x *plannercore.PhysicalSelection) {
+	estRows := plannercore.GetPhysicalPlanEstRowCount(x)
+	actualRows := GetResultRowsCountForCopStats(stmtCtx, x.ID())
```

Comment:
This function can fail (-1 return). Can you also bail out in that case?

Reply:
ok.